go/enclave/storage/events_storage.go

```diff
@@ -27,122 +27,197 @@ func newEventsStorage(cachingService *CacheService, logger gethlog.Logger) *even
+}
+
+func (es *eventsStorage) storeNewContract(ctx context.Context, dbTX *sql.Tx, createdContract gethcommon.Address, senderId *uint64, cfg *core.ContractVisibilityConfig) (*uint64, error) {
+	ctrId, err := enclavedb.WriteContractConfig(ctx, dbTX, createdContract, *senderId, cfg)
+	if err != nil {
+		return nil, fmt.Errorf("could not write contract address. cause %w", err)
+	}
+	return ctrId, nil
+}
+
 func (es *eventsStorage) storeEventLog(ctx context.Context, dbTX *sql.Tx, execTxId uint64, l *types.Log) error {
-	topicIds, isLifecycle, err := es.handleUserTopics(ctx, dbTX, l)
+	eventSig := l.Topics[0]
+
+	contract, err := es.readContract(ctx, dbTX, l.Address)
 	if err != nil {
-		return err
+		// the contract should already have been stored when it was created
+		return fmt.Errorf("could not read contract address. %s. Cause: %w", l.Address, err)
 	}
 
-	eventTypeId, err := es.handleEventType(ctx, dbTX, l, isLifecycle)
+	eventType, err := es.readEventType(ctx, dbTX, l.Address, eventSig)
+	if errors.Is(err, errutil.ErrNotFound) {
+		// this is the first type an event of this type is emitted, so we must store it
+		eventType, err = es.storeEventType(ctx, dbTX, contract, l)
+		if err != nil {
+			return fmt.Errorf("could not write event type. cause %w", err)
+		}
+	} else if err != nil {
+		// unexpected event type
+		return fmt.Errorf("could not read event type. Cause: %w", err)
+	}
+
+	topicIds, err := es.storeTopics(ctx, dbTX, l)
 	if err != nil {
-		return err
+		return fmt.Errorf("could not store topics. cause: %w", err)
 	}
 
 	// normalize data
 	data := l.Data
 	if len(data) == 0 {
 		data = nil
 	}
-	err = enclavedb.WriteEventLog(ctx, dbTX, eventTypeId, topicIds, data, l.Index, execTxId)
+	err = enclavedb.WriteEventLog(ctx, dbTX, eventType.Id, topicIds, data, l.Index, execTxId)
 	if err != nil {
 		return fmt.Errorf("could not write event log. Cause: %w", err)
 	}
 
 	return nil
 }
 
-func (es *eventsStorage) handleEventType(ctx context.Context, dbTX *sql.Tx, l *types.Log, isLifecycle bool) (uint64, error) {
-	et, err := es.readEventType(ctx, dbTX, l.Address, l.Topics[0])
-	if err != nil && !errors.Is(err, errutil.ErrNotFound) {
-		return 0, fmt.Errorf("could not read event type. Cause: %w", err)
-	}
-	if err == nil {
-		// in case we determined the current emitted event is not lifecycle, we must update the EventType
-		if !isLifecycle && et.isLifecycle {
-			err := enclavedb.UpdateEventTypeLifecycle(ctx, dbTX, et.id, isLifecycle)
-			if err != nil {
-				return 0, fmt.Errorf("could not update the event type. cause: %w", err)
-			}
+// handles the visibility config detection
+func (es *eventsStorage) storeEventType(ctx context.Context, dbTX *sql.Tx, contract *enclavedb.Contract, l *types.Log) (*enclavedb.EventType, error) {
+	eventType := enclavedb.EventType{ContractId: contract.Id, EventSignature: l.Topics[0], AutoVisibility: contract.AutoVisibility}
+
+	// when the contract is transparent, all events are public
+	switch {
```

Comment:
Feels a bit confusing to me to use a switch here instead of an if-else for these sorts of conditions.